File: Projects/Project_04_1D_SWE/Old/For_Students/matlab/global_matrices.m
%---------------------------------------------------------------------%
%This function computes the LGL grid and elements.
%Written by F.X. Giraldo on 10/2003
%           Department of Applied Mathematics
%           Naval Postgraduate School
%           Monterey, CA 93943-5216
%---------------------------------------------------------------------%
function [M,D,F] = global_matrices(Me,De,Fe,intma,coord,npoin,nelem,ngl,periodicity)

%Initialize
M=zeros(npoin,npoin);
D=zeros(npoin,npoin);
F=zeros(npoin,npoin);

%Form Global Mass and Differentiation Matrices
for e=1:nelem
    for i=1:ngl
        I=intma(i,e);
        x(i)=coord(I);
        inode(i)=periodicity(I);
    end
    jac=0.5*(x(ngl)-x(1));
    for j=1:ngl
        J=inode(j);
        for i=1:ngl
            I=inode(i);
            M(I,J)=M(I,J) + jac*Me(i,j);
            D(I,J)=D(I,J) + De(i,j);
            F(I,J)=F(I,J) + Fe(i,j);
        end
    end
end 

if periodicity(npoin) == periodicity(1)
    M(npoin,npoin)=1;
end

File: Projects/Project_03b_1D_Burger_Entropy_Stable/For_Instructors/julia/global_matrices.jl
#---------------------------------------------------------------------#
#This function computes the global Mass and Differentiation Matrices.
#Written by F.X. Giraldo on April 19 2019
#           Department of Applied Mathematics
#           Naval Postgraduate School
#           Monterey; CA 93943-5216
#---------------------------------------------------------------------#
function global_matrices(Me,De,Dwe,Fe,intma,coord,Npoin,Ne,Np,periodicity,DFloat)

    #Initialize
    M=zeros(DFloat,Npoin,Npoin)
    D=zeros(DFloat,Npoin,Npoin)
    Dw=zeros(DFloat,Npoin,Npoin)
    F=zeros(DFloat,Npoin,Npoin)
    inode=zeros(Int64,Np)
    x=zeros(DFloat,Np)

    #Loop through elements
    for e=1:Ne
        #Store Coordinates
        for i=1:Np
            I=intma[i,e]
            inode[i]=periodicity[I]
            x[i]=coord[I]
        end

        #Element Length & Jacobian
        dx=x[Np]-x[1]
        jac=dx/2

        #Perform DSS
        for j=1:Np
            J=inode[j]
            for i=1:Np
                I=inode[i]
                M[I,J]+=jac*Me[i,j]
                D[I,J]+=De[i,j]
                Dw[I,J]+=Dwe[i,j]
                F[I,J]+=Fe[i,j]
            end #j
        end #i
    end #e

    #Periodicity
    if periodicity[Npoin] == periodicity[1]
        M[Npoin,Npoin]=1
    end

    #check Fmatrix
#     for i=1:Npoin
#         for j=1:Npoin
#             println("i = ",i," j = ",j," F = ",F[i,j])
#         end
#     end

    #Return arrays
    return(M,D,Dw,F)
end
